apps/server/src/filter/jwt-auth.guard.ts

```diff
@@ -0,0 +1,123 @@
+   */
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    if (
+      this.reflector.getAllAndOverride<boolean>(SKIP_AUTH_KEY, [
+        context.getHandler(),
+        context.getClass(),
+      ])
+    ) {
+      return true;
+    }
+
+    // Proceed with standard authentication check
+    if (!(await super.canActivate(context))) {
+      return false;
+    }
+
+    // Retrieve roles from controller and method context
+    const roles = this.getRoles(context);
+
+    // If no roles are specified, allow access
+    if (!roles.length) return true;
+
+    // Get user from request and verify role access
+    const userRoles = context.switchToHttp().getRequest().use.roles;
+    if (this.hasRequiredRoles(userRoles, roles)) {
+      return true;
+    }
+
+    throw new UnauthorizedException('User does not have the required roles');
+  }
+
+  /**
+   * Combines controller and method roles into a single array.
+   * @param {ExecutionContext} context - The execution context of the route.
+   * @returns {string[]} - The combined roles.
+   */
+  getRoles(context: ExecutionContext): string[] {
+    const controllerRoles = this.reflector.getAllAndOverride<string[]>(
+      ROLES_KEY,
+      [context.getClass()],
+    );
+    const methodRoles = this.reflector.getAllAndOverride<string[]>(ROLES_KEY, [
+      context.getHandler(),
+    ]);
+
+    // Combine and deduplicate roles
+    return [...new Set([...(controllerRoles || []), ...(methodRoles || [])])];
+  }
+
+  /**
+   * Checks if the user has at least one of the required roles.
+   * @param {RoleEnum[]} userRoles - The user object from the request.
+   * @param {string[]} roles - The required roles.
+   * @returns {boolean} - Whether the user has required roles.
+   */
+  hasRequiredRoles(userRoles: RoleEnum[], roles: string[]): boolean {
+    return userRoles?.some((userRole) => roles.includes(userRole));
+  }
+
+  /**
+   * Handles the result of the authentication.
+   * @param {Error} err - The error object.
+   * @param {User} user - The user object, if authentication is successful.
+   * @param {string} info - Additional information about the authentication process.
+   * @returns {User} - The user object if authentication is successful.
+   * @throws {RestException} - Throws a `RestException` if authentication fails.
+   */
+  handleRequest<TUser = User>(
+    err: Error | null,
+    user: User | null,
+    info: string,
+  ): TUser {
+    if (err || !user) {
+      const unauthorizedException = new UnauthorizedException();
+      const errorMessage = new ErrorMessage(
+        unauthorizedException.getStatus(),
+        unauthorizedException.message.concat(':').concat(info), //will remove concat block after discussion regarding if we need info at all
+        ErrorDescription.UNAUTHORIZED_USER,
+      );
+      throw new RestException(errorMessage);
+    }
+
+    return user as unknown as TUser; //unsure how  not to use the return type as any here.
```

Comment:
If we're aiming to return the user of type User, this doesn't satisfy condition enough to ensure user is of type User.

Reply:
Will create ticket  to resolve this in next PR